borgmatic/commands/completion.py

```diff
@@ -55,3 +65,153 @@ def bash_completion():
             '\ncomplete -o bashdefault -o default -F complete_borgmatic borgmatic',
         )
     )
+
+
+# fish section
+
+
+def has_file_options(action: Action):
+    return action.metavar in (
+        'FILENAME',
+        'PATH',
+    ) or action.dest in ('config_paths',)
+
+
+def has_choice_options(action: Action):
+    return action.choices is not None
+
+
+def has_required_param_options(action: Action):
+    return (
+        action.required is True
+        or action.nargs
+        in (
+            '+',
+            '*',
+        )
+        or '--archive' in action.option_strings
```

Comment:
I noticed that above you do `action.dest in ('config_paths',)`, but here you check the `option_strings`. However it might be nice to pick one approach and do it everywhere you're trying to identify a particular command-line flag.

Do not feel strongly.

Reply:
Also, what's the purpose of this particular check? Just to match the couple of actions which have an optional rather than required `--archive` flag? But then isn't that kind of violating the name of the function.. `has_required_param_options()`? (This is definitely a case where a docstring for the function would help me out!)

Reply:
The purpose is to prevent the normal completions from showing up when you hit tab. Given the default is already set for archive, its sorta weird to use it without specifying an archive... but I removed it. Much more helpful if the completions could know about archives. Could borgmatic perhaps have a method that returns a cached list of archives from the last time it interacted with the server, that doesn't require auth?

Reply:
Thanks for the explanation! A cached list of archives would be possible—but probably not as part of this initial ticket. Feel free to file a new one for that feature though. It might be a little tricky because borgmatic doesn't (currently) list archives upon create. Although I guess it could implicitly append any newly created archive names to the cache without having to run `list`...